src/LogisticSparseTaskS3.cpp

```diff
@@ -91,6 +91,21 @@ void LogisticSparseTaskS3::run(const Configuration& config, int worker) {
   uint64_t version = 1;
   SparseLRModel model(1 << config.get_model_bits());
 
+  if (ps_ips.size() == 1) {
+    psint = std::make_unique<PSSparseServerInterface>(ps_ips[0], ps_ports[0]);
+  } else {
+    psint = std::make_unique<MultiplePSSparseServerInterface>(config, ps_ips,
+                                                              ps_ports);
+  }
+  while (true) {
```

Comment:
You can abstract this while loop (used in other places).

Like this.

```
class A { 
  public:
    void connect() {
      std::cout << "test" << std::endl;
    }   
};

void repeat(const std::function<void()>& f) {
  while (true) {
    try {
      f();
      break;
    } catch (const std::exception& exc) {
      std::cout << exc.what();
    }   
  }
}

int main() {
  A a;
  repeat(std::bind(&A::connect, &a));
  return 0;
}
```

Reply:
Fixed